File: sdk/basin-js/src/storage.ts
/**
 * `.storage` — bindings for `/storage/v1/*` (ADR 0021 surface).
 *
 * Route sources (verified, crates/basin-rest/src/server.rs:373-424; handlers
 * in crates/basin-rest/src/routes/storage.rs and storage_sign.rs):
 *
 * - `POST   /storage/v1/bucket`                                  create bucket
 * - `GET    /storage/v1/bucket/:name`                            bucket metadata
 * - `DELETE /storage/v1/bucket/:name`                            delete + purge
 * - `POST   /storage/v1/object/:bucket/*path`                    upload (JWT)
 * - `GET    /storage/v1/object/:bucket/*path`                    download (JWT)
 * - `DELETE /storage/v1/object/:bucket/*path`                    delete (JWT)
 * - `POST   /storage/v1/object/list/:bucket`                     list (body: prefix/limit/offset)
 * - `DELETE /storage/v1/object/:bucket`                          bulk delete (body: prefixes[])
 * - `GET    /storage/v1/object/public/:project_id/:bucket/*path` public download (no JWT)
 * - `POST   /storage/v1/object/sign/upload/:bucket/*path`        mint signed URL (JWT)
 * - `GET    /storage/v1/object/sign/:project/:bucket/*path`      signed download (no JWT)
 */

import { BasinApiError } from "./errors.js";
import type { ClientContext } from "./http.js";
import { encodeObjectPath, requestJson, requestRaw } from "./http.js";
import type { Bucket, SignedUrl, StorageObject } from "./types.js";

export interface CreateBucketOptions {
  public?: boolean;
  fileSizeLimit?: number;
  allowedMimeTypes?: string[];
}

export interface DownloadResult {
  data: ArrayBuffer;
  contentType: string | null;
  etag: string | null;
}

export class StorageClient {
  #ctx: ClientContext;

  constructor(ctx: ClientContext) {
    this.#ctx = ctx;
  }

  /** `POST /storage/v1/bucket` (201). */
  async createBucket(
    name: string,
    opts: CreateBucketOptions = {},
  ): Promise<Bucket> {
    return requestJson(this.#ctx, "POST", "/storage/v1/bucket", {
      body: {
        name,
        public: opts.public ?? false,
        file_size_limit: opts.fileSizeLimit ?? null,
        allowed_mime_types: opts.allowedMimeTypes ?? [],
      },
    });
  }

  /** `GET /storage/v1/bucket/:name`. */
  async getBucket(name: string): Promise<Bucket> {
    return requestJson(
      this.#ctx,
      "GET",
      `/storage/v1/bucket/${encodeURIComponent(name)}`,
    );
  }

  /** `DELETE /storage/v1/bucket/:name` — also purges object bytes. */
  async deleteBucket(
    name: string,
  ): Promise<{ message: string; objects_purged: number }> {
    return requestJson(
      this.#ctx,
      "DELETE",
      `/storage/v1/bucket/${encodeURIComponent(name)}`,
    );
  }

  /** Object operations scoped to one bucket. */
  from(bucket: string): StorageBucketClient {
    return new StorageBucketClient(this.#ctx, bucket);
  }
}

export class StorageBucketClient {
  #ctx: ClientContext;
  #bucket: string;

  constructor(ctx: ClientContext, bucket: string) {
    this.#ctx = ctx;
    this.#bucket = bucket;
  }

  #objectPath(path: string): string {
    return `/storage/v1/object/${encodeURIComponent(this.#bucket)}/${encodeObjectPath(path)}`;
  }

  /**
   * `POST /storage/v1/object/:bucket/*path` — upload. MIME is taken from the
   * `Content-Type` header (extensions are never trusted server-side).
   */
  async upload(
    path: string,
    body: BodyInit,
    opts: { contentType?: string } = {},
  ): Promise<StorageObject> {
    const headers: Record<string, string> = {};
    if (opts.contentType !== undefined) headers["content-type"] = opts.contentType;
    return requestJson(this.#ctx, "POST", this.#objectPath(path), {
      rawBody: body,
      headers,
    });
  }

  /** `GET /storage/v1/object/:bucket/*path` — authenticated download. */
  async download(path: string): Promise<DownloadResult> {
    const res = await requestRaw(this.#ctx, "GET", this.#objectPath(path));
    return {
      data: await res.arrayBuffer(),
      contentType: res.headers.get("content-type"),
      etag: res.headers.get("etag"),
    };
  }

  /** `DELETE /storage/v1/object/:bucket/*path` — single delete. */
  async remove(path: string): Promise<{ message: string }> {
    return requestJson(this.#ctx, "DELETE", this.#objectPath(path));
  }

  /** `POST /storage/v1/object/list/:bucket` — list with optional prefix. */
  async list(
    opts: { prefix?: string; limit?: number; offset?: number } = {},
  ): Promise<StorageObject[]> {
    return requestJson(
      this.#ctx,
      "POST",
      `/storage/v1/object/list/${encodeURIComponent(this.#bucket)}`,
      {
        body: {
          prefix: opts.prefix ?? null,
          limit: opts.limit ?? null,
          offset: opts.offset ?? null,
        },
      },
    );
  }

  /**
   * `DELETE /storage/v1/object/:bucket` with `{ prefixes: [...] }` —
   * bulk delete every object matching any prefix. RLS-denied objects are
   * skipped server-side (not counted in `deleted`).
   */
  async removeByPrefixes(prefixes: string[]): Promise<{ deleted: number }> {
    return requestJson(
      this.#ctx,
      "DELETE",
      `/storage/v1/object/${encodeURIComponent(this.#bucket)}`,
      { body: { prefixes } },
    );
  }

  /**
   * Build the no-JWT public URL:
   * `GET /storage/v1/object/public/:project_id/:bucket/*path`.
   * Requires a known project id; only works for buckets with `public: true`.
   */
  getPublicUrl(path: string): string {
    const project = this.#ctx.projectId();
    if (project === undefined) {
      throw new BasinApiError(
        "E_INVALID_REQUEST",
        "public URLs require a project id (createClient options.projectId or a JWT key/session)",
        0,
      );
    }
    return `${this.#ctx.baseUrl}/storage/v1/object/public/${project}/${encodeURIComponent(this.#bucket)}/${encodeObjectPath(path)}`;
  }

  /**
   * `POST /storage/v1/object/sign/upload/:bucket/*path` — mint a time-boxed
   * signed *download* URL (despite the `upload` literal, which only
   * disambiguates the axum route — see storage_sign.rs). TTL is capped at 7
   * days server-side; default 3600s.
   */
  async createSignedUrl(
    path: string,
    opts: { expiresIn?: number } = {},
  ): Promise<SignedUrl & { absoluteUrl: string }> {
    const signed = await requestJson<SignedUrl>(
      this.#ctx,
      "POST",
      `/storage/v1/object/sign/upload/${encodeURIComponent(this.#bucket)}/${encodeObjectPath(path)}`,
      { body: { expires_in: opts.expiresIn ?? 3600 } },
    );
    return { ...signed, absoluteUrl: this.#ctx.baseUrl + signed.signedUrl };
  }
}

File: sdk/basin-dotnet/src/Basin/Storage.cs
using System.Text.Json;

namespace Basin;

/// <summary>
/// Storage client for /storage/v1/* routes.
///
/// Route sources (verified against crates/basin-rest/src/server.rs and
/// crates/basin-rest/src/routes/storage.rs + storage_sign.rs):
/// - POST   /storage/v1/bucket                                  create bucket
/// - GET    /storage/v1/bucket/:name                            bucket metadata
/// - DELETE /storage/v1/bucket/:name                            delete + purge
/// - POST   /storage/v1/object/:bucket/*path                    upload (JWT)
/// - GET    /storage/v1/object/:bucket/*path                    download (JWT)
/// - DELETE /storage/v1/object/:bucket/*path                    delete (JWT)
/// - POST   /storage/v1/object/list/:bucket                     list (body: prefix/limit/offset)
/// - DELETE /storage/v1/object/:bucket                          bulk delete (body: prefixes[])
/// - GET    /storage/v1/object/public/:project_id/:bucket/*path public download (no JWT)
/// - POST   /storage/v1/object/sign/upload/:bucket/*path        mint signed URL (JWT)
/// - GET    /storage/v1/object/sign/:project/:bucket/*path      signed download (no JWT)
/// </summary>
public sealed class StorageClient
{
    private readonly BasinTransport _transport;
    private readonly Func<string?> _projectIdGetter;

    internal StorageClient(BasinTransport transport, Func<string?> projectIdGetter)
    {
        _transport = transport;
        _projectIdGetter = projectIdGetter;
    }

    /// <summary>POST /storage/v1/bucket (201).</summary>
    public async Task<Bucket> CreateBucketAsync(
        string name,
        bool @public = false,
        long? fileSizeLimit = null,
        IEnumerable<string>? allowedMimeTypes = null,
        CancellationToken ct = default)
    {
        var result = await _transport.RequestJsonAsync<Bucket>(
            HttpMethod.Post,
            "/storage/v1/bucket",
            body: new
            {
                name,
                @public,
                file_size_limit = fileSizeLimit,
                allowed_mime_types = (allowedMimeTypes ?? Enumerable.Empty<string>()).ToArray()
            },
            ct: ct).ConfigureAwait(false);

        return result!;
    }

    /// <summary>GET /storage/v1/bucket/:name.</summary>
    public async Task<Bucket> GetBucketAsync(string name, CancellationToken ct = default)
    {
        var result = await _transport.RequestJsonAsync<Bucket>(
            HttpMethod.Get,
            $"/storage/v1/bucket/{Uri.EscapeDataString(name)}",
            ct: ct).ConfigureAwait(false);

        return result!;
    }

    /// <summary>DELETE /storage/v1/bucket/:name — also purges object bytes.</summary>
    public Task<JsonElement?> DeleteBucketAsync(string name, CancellationToken ct = default) =>
        _transport.RequestJsonAsync<JsonElement?>(
            HttpMethod.Delete,
            $"/storage/v1/bucket/{Uri.EscapeDataString(name)}",
            ct: ct);

    /// <summary>Return an object-level client scoped to one bucket.</summary>
    public StorageBucketClient FromBucket(string bucket) =>
        new(_transport, bucket, _projectIdGetter);
}

/// <summary>Object operations scoped to one bucket.</summary>
public sealed class StorageBucketClient
{
    private readonly BasinTransport _transport;
    private readonly string _bucket;
    private readonly Func<string?> _projectIdGetter;

    internal StorageBucketClient(
        BasinTransport transport,
        string bucket,
        Func<string?> projectIdGetter)
    {
        _transport = transport;
        _bucket = bucket;
        _projectIdGetter = projectIdGetter;
    }

    private string ObjectPath(string path) =>
        $"/storage/v1/object/{Uri.EscapeDataString(_bucket)}/{EncodeObjectPath(path)}";

    private static string EncodeObjectPath(string path) =>
        string.Join("/", path.Split('/').Select(Uri.EscapeDataString));

    /// <summary>POST /storage/v1/object/:bucket/*path — upload bytes.</summary>
    public async Task<StorageObject> UploadAsync(
        string path,
        byte[] data,
        string? contentType = null,
        CancellationToken ct = default)
    {
        var extraHeaders = contentType is not null
            ? new Dictionary<string, string> { ["Content-Type"] = contentType }
            : null;

        // Storage upload uses raw bytes, not JSON encoding.
        var resp = await _transport.RequestAsync(
            HttpMethod.Post,
            ObjectPath(path),
            body: data,
            contentType: contentType ?? "application/octet-stream",
            extraHeaders: extraHeaders,
            ct: ct).ConfigureAwait(false);

        var raw = await resp.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        return JsonSerializer.Deserialize<StorageObject>(raw, BasinJsonOptions.Default)!;
    }

    /// <summary>GET /storage/v1/object/:bucket/*path — authenticated download.</summary>
    public async Task<DownloadResult> DownloadAsync(string path, CancellationToken ct = default)
    {
        var (data, contentType, etag) = await _transport.RequestBytesAsync(
            HttpMethod.Get,
            ObjectPath(path),
            ct: ct).ConfigureAwait(false);

        return new DownloadResult { Data = data, ContentType = contentType, ETag = etag };
    }

    /// <summary>DELETE /storage/v1/object/:bucket/*path — single delete.</summary>
    public Task<JsonElement?> RemoveAsync(string path, CancellationToken ct = default) =>
        _transport.RequestJsonAsync<JsonElement?>(
            HttpMethod.Delete,
            ObjectPath(path),
            ct: ct);

    /// <summary>POST /storage/v1/object/list/:bucket.</summary>
    public async Task<IReadOnlyList<StorageObject>> ListAsync(
        string? prefix = null,
        int? limit = null,
        int? offset = null,
        CancellationToken ct = default)
    {
        var result = await _transport.RequestJsonAsync<List<StorageObject>>(
            HttpMethod.Post,
            $"/storage/v1/object/list/{Uri.EscapeDataString(_bucket)}",
            body: new { prefix, limit, offset },
            ct: ct).ConfigureAwait(false);

        return result ?? new List<StorageObject>();
    }

    /// <summary>
    /// DELETE /storage/v1/object/:bucket with { prefixes: [...] }.
    /// Bulk-delete every object matching any prefix.
    /// </summary>
    public Task<JsonElement?> RemoveByPrefixesAsync(
        IEnumerable<string> prefixes,
        CancellationToken ct = default) =>
        _transport.RequestJsonAsync<JsonElement?>(
            HttpMethod.Delete,
            $"/storage/v1/object/{Uri.EscapeDataString(_bucket)}",
            body: new { prefixes = prefixes.ToArray() },
            ct: ct);

    /// <summary>
    /// Build the no-JWT public URL.
    /// GET /storage/v1/object/public/:project_id/:bucket/*path
    /// Only works for buckets with Public=true.
    /// </summary>
    public string GetPublicUrl(string path)
    {
        var project = _projectIdGetter();
        if (project is null)
            throw new BasinApiException(
                "E_INVALID_REQUEST",
                "public URLs require a project id — pass ProjectId to BasinClient.Builder",
                0);

        return $"{_transport.BaseUrl}/storage/v1/object/public/{project}" +
               $"/{Uri.EscapeDataString(_bucket)}/{EncodeObjectPath(path)}";
    }

    /// <summary>
    /// POST /storage/v1/object/sign/upload/:bucket/*path — mint a signed URL.
    /// TTL is capped at 7 days server-side; default 3600 s.
    /// The signed URL is for download despite "upload" in the path (the segment
    /// only disambiguates the axum router — see storage_sign.rs).
    /// </summary>
    public async Task<SignedUrl> CreateSignedUrlAsync(
        string path,
        int expiresIn = 3600,
        CancellationToken ct = default)
    {
        var resp = await _transport.RequestAsync(
            HttpMethod.Post,
            $"/storage/v1/object/sign/upload/{Uri.EscapeDataString(_bucket)}/{EncodeObjectPath(path)}",
            body: JsonSerializer.SerializeToUtf8Bytes(new { expires_in = expiresIn }, BasinJsonOptions.Default),
            contentType: "application/json",
            ct: ct).ConfigureAwait(false);

        var raw = await resp.Content.ReadAsStringAsync(ct).ConfigureAwait(false);

        // Wire fields are camelCase: signedUrl / expiresAt (storage_sign.rs).
        using var doc = JsonDocument.Parse(raw);
        var root = doc.RootElement;
        var rel = root.TryGetProperty("signedUrl", out var su) ? su.GetString() ?? "" : "";
        var expiresAt = root.TryGetProperty("expiresAt", out var ea) ? ea.GetString() ?? "" : "";

        return new SignedUrl
        {
            SignedUrlPath = rel,
            ExpiresAt = expiresAt,
            AbsoluteUrl = _transport.BaseUrl + rel,
        };
    }
}
